### DISCUSSION TITLE
Error to do with low duration runs with continue_run=True

### DISCUSSION THREAD

**Discussion by HenryMarsden** [neurolib-dev/neurolib#196] [Q&A] [answered]
Hi guys,

I would like to start by saying thank you very much for this software, it has very interesting exploring it for my dissertation.

Anyway, I've found an error that occurs when the duration of the simulation is set too low (>100ms). This error is strange as it isn't a binary thing, instead the probability of the error increases as the duration decreases. I do not think the error is linked to the continue_run parameter. The error seems to occur at durations as high as 250ms with about a 20% probability from my observations. One detail that may help is that the failed runs tend to occur in pulses as do the successful runs, in other words they come in clumps.

Please let me know if you get any ideas as to how I could get around this error as I would really like to try some tests that require a greater degree of precision.

Thank you,
Henry Marsden

PS, here is my code
![thumbnail_CodeScreenGrab](https://user-images.githubusercontent.com/59068823/161264639-a5d83d47-781f-47b8-8b19-f91e7bf44158.jpg)

**Comment by jajcayn** [comment]
Hey @HenryMarsden,

thanks for giving a shot to `neurolib`:) I cannot say just by a look what might be wrong. Thanks for the code, I'll try to debug later today:) Will be in touch,

Best,
N.

**Reply by HenryMarsden** [reply]
Thanks for the quick reply!

I forgot to put the actual error code too so I will leave it here

'ERROR:root:nan in model output!'

**Reply by jajcayn** [reply]
I just realised that your code is a screenshot:) Could you please post your code using copy-paste into a code block? (Code block on github can be done using three `).
It's much easier to copy than write from screenshot:)
Cheers

**Reply by HenryMarsden** [reply]
Sorry about that, here you go!
```
import matplotlib.pyplot as plt

from neurolib.models.thalamus import ThalamicMassModel

thalamus = ThalamicMassModel()

## This parameter seems to be what causes the error in conjunction with 'continue_run=True'
duration = 10


thalamus.params['duration'] = duration

thalamus.run(continue_run=True)

plt.plot(thalamus['t'], thalamus['Q_t'].T, lw=2, c="r",)
plt.title('Q_t 1hz stimulus 20ms width')
plt.xlabel("Time (ms)")
plt.ylabel("Coulombs (C)")

thalamus.params['duration'] = duration

for x in range(50):
    thalamus.run(continue_run=True)
    plt.plot((thalamus['t'] + ((x+1) * duration)), thalamus['Q_t'].T, lw=2, c="r",)
```

**Reply by HenryMarsden** [reply]
I changed the code to be more simple but still containing the bug so that its easier to debug as well

Thanks for the help

**Comment by caglorithm** [comment]
Hi Henry,

You told me earlier that you have the error `ERROR:root:nan in model output!`. This usually means that the “model blew up” or something else caused `nan`’s to appear in the time series (which should not happen if everything goes right). 

Maybe your stimulus is too strong? The input `stimulus = 2` needs to be adjusted such that it is in the same order of the model's variables.

**Reply by HenryMarsden** [reply]
Hi @caglorithm,

Thank you for the response. I have tried removing the stimulus and the error persists (The code I put in one of my other replies still get the error and has to external stimulus) so I don't think that is the issue.

**Comment by jajcayn** [comment, marked as answer]
So @caglorithm is right.
I just tested your code with lower `dt`: when you lower your dt the model runs smoothly, try e.g. dt = 0.001 ms using
```python
thalamus.params['duration'] = duration
thalamus.params['dt'] = 0.001
```

Should work:)

**Reply by HenryMarsden** [reply]
Thank you so much!

I don't know why that parameter completely slipped my mind. This is incredibly helpful, Thank you!

**Reply by HenryMarsden** [reply]
Sorry to be a bother I just had one quick question.

I was just looking through the thalamus default parameters and found this comment 
```
### runtime parameters
    # thalamus is really sensitive, so either you integrate with very small dt or use an adaptive integration step
    params.dt = 0.1  # ms //This is the lowest dt possible due to memory constraints //Old 0.01
```
should I be worried about this? and if not how low is it possible to go, as I would ideally like to set duration as low as 10ms.

Thanks

**Reply by jajcayn** [reply]
Hey @HenryMarsden, so I tried with dt=0.01 and that works as well...
note about memory: so you need to take into account the total duration of the run + sampling, not only sampling...
e.g. with 10 seconds run and dt=0.1 ms you have 100 000 samples of float64.. thalamus is a pretty complicated model with approx. 25 variables so each of the variables has 100 000 samples... this is probably fine on a modern computer. However, consider doing 2 minute run with 0.001 ms dt, now you have 120 000 000 samples per variable...

Summed: for short runs, you can go to lower dt without problems.. for longer runs, this might get tricky...

BTW, I tested the thalamus with dt=0.1 and continue_run without stimulus and it was working nicely, so you need to think a little bit more about the stimulus and its size

**Reply by HenryMarsden** [reply]
What you said about memory is very interesting, I will have to take that into account as I just crashed my computer.

I tried changing the dt and for me it only works if I go as low as 0.00001, do you have any idea why it may differ on my computer? (It is a laptop so not the most powerful) As when I try the values you used the error persists.
Thanks

**Reply by jajcayn** [reply]
Try to always run it multiple times in the same session. First, go low with your dt, e.g. dt=0.001. Then for a subsequent run you can make your dt higher - I tested it, first run in jupyter with dt=0.001 and then even dt=0.1 worked nicely.
I am not sure about the culprit, but in the background, we are using `numba` compilation into a machine code to speed everything up. The first run within one session is a "compilation run" and then everything is ready to use. I guess for some cryptic reason, the first, ergo compilation run has to be with low dt.
